lib/client/index.js

```diff
@@ -196,17 +196,20 @@ class Client {
      * Return a url that can be shared by others to acess a file.
      *
      * @param {string} path
+     * @param {object} [opts]
+     * @param {boolean} [opts.encrypted]
      *
      * @returns {Promise<string>}
      */
-  async createURL (path) {
+  async createURL (path, opts = {}) {
     const fullPath = this._fullPath(path)
     const record = await this._getStoredRecord(fullPath)
 
     const query = `${this._relay ? 'relay=' + this._relay : ''}`
     const fragment = {}
 
-    if (record?.metadata?.encrypted) {
+    // If the file is encrypted or we want to get url with key to file that does not exist yet
+    if (record?.metadata?.encrypted || (!record && opts.encrypted )) {
```

Comment:
Small point, but perhaps `opts.encrypt` instead to be consistent with `put()` method.

Reply:
Then for completeness:
```diff
@@ -85,10 +85,9 @@ class Client {
    */
   async put (path, content, opts = {}) {
     const fullPath = this._fullPath(path)
+    const parsed = this._parseURL(path)
 
-    content = opts.encrypt ? await this._encrypt(fullPath, content) : content
+    content = (opts.encrypt || parsed?.encryptionKey) ? await this._encrypt(fullPath, content, parsed?.encryptionKey) : content
 
     const record = await Record.create(this._keyPair, fullPath, content, { metadata: { encrypted: opts.encrypt } })
 
@@ -554,11 +553,12 @@ class Client {
   /**
    * @param {string} path
    * @param {Uint8Array} content
+   * @param {Uint8Array} key
    */
-  async _encrypt (path, content) {
+  async _encrypt (path, content, key) {
     if (content.length === 0) return content
 
-    const encryptionKey = await this._generateEncryptionKey(path)
+    const encryptionKey = key ?? await this._generateEncryptionKey(path)
 
     // <nonce><authentication tag><encrypted content>
     const result = Buffer.alloc(sodium.crypto_secretbox_NONCEBYTES + sodium.crypto_secretbox_MACBYTES + content.length)
```

Which will allow `put()` using same uri and you'll have a r/w 'shareable' encrypted file.

Reply:
Good point. Chaning. I was on the edge due to possible confusion that returned url itself can be encrypted

Reply:
Actually scratch my second comment. This would be a bigger change as `Record` signatures won't match when `put`ing to remote relay without key delegation.

Reply:
> when puting to remote relay without key delegation.

not sure I understand what do you mean by "without key delegation." and thus why signatures will not match

Reply:
Let's say you try this at the end of the same test you modified:
```js
  const urlWithEncryptKey = await a.createURL('bar', { encrypt: true })
  await b.put(urlWithEncryptKey, value, { awaitRelaySync: true })
  t.alike(await a.get('bar'), value, 'get shared encrypted file (e2e)')
  // throws Error: Invalid signature
```
`b` **signs** and puts the record, then `a` attempts to verify on get using (its own) z32 id in the url as pubkey => sig mismatch.

One workaround would be to include signer `b`'s pubkey in the encoded url as query param, ie. `{ shared: [ key: <z32 id string>, expiry?: <timestamp> ] }`. When put, the `shared` obj array is saved with the record metadata, then peers, ie. `a`, could verify the signature against those pubkeys provided the expiry (if specified) has not elapsed.

However, this is not secure as it allows publishing arbitrary data to a peer's relay & id.

Easier: `a` creates a signed delegate record/blob for `b`'s pubkey + expiry beforehand (similar to [NIP26](https://github.com/nostr-protocol/nips/blob/master/26.md)).

Harder: 3-round [MuSig](https://eprint.iacr.org/2018/068).

Reply:
ok, it makes sense that `b` should not be able to write to `a`'s file. I've misunderstood the intention of [your comment](https://github.com/slashtags/web-relay/pull/52#discussion_r1415712319)  then. I thought it was for passing encryption key which can be created in by user in the way they want. Made me wonder why not pass it in opts then... Reverting